src/ML-PACE/compute_pace.cpp

```diff
@@ -0,0 +1,526 @@
+// clang-format off
+/* ----------------------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://www.lammps.org/, Sandia National Laboratories
+   LAMMPS Development team: developers@lammps.org
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+#include "compute_pace.h"
+#include "ace-evaluator/ace_evaluator.h"
+#include "ace-evaluator/ace_c_basis.h"
+#include "ace-evaluator/ace_abstract_basis.h"
+#include "ace-evaluator/ace_types.h"
+#include <cstring>
+#include <map>
+
+#include "atom.h"
+#include "update.h"
+#include "modify.h"
+#include "neighbor.h"
+#include "neigh_list.h"
+#include "neigh_request.h"
+#include "force.h"
+#include "pair.h"
+#include "comm.h"
+#include "memory.h"
+#include "error.h"
+
+namespace LAMMPS_NS {
+struct ACECimpl {
+  ACECimpl() : basis_set(nullptr), ace(nullptr) {}
+  ~ACECimpl()
+  {
+    delete basis_set;
+    delete ace;
+  }
+  ACECTildeBasisSet *basis_set;
+  ACECTildeEvaluator *ace;
+};
+}
+
+using namespace LAMMPS_NS;
+
+enum{SCALAR,VECTOR,ARRAY};
+ComputePACE::ComputePACE(LAMMPS *lmp, int narg, char **arg) :
+  Compute(lmp, narg, arg), cutsq(nullptr), list(nullptr), pace(nullptr),
+  paceall(nullptr), pace_peratom(nullptr), map(nullptr), cg(nullptr)
+{
+  array_flag = 1;
+  extarray = 0;
+  bikflag = 0;
+  dgradflag = 0;
+
+  int ntypes = atom->ntypes;
+  int nargmin = 4;
+
+  acecimpl = new ACECimpl;
+  if (narg < nargmin) error->all(FLERR,"Illegal compute pace command");
+
+  bikflag = utils::inumeric(FLERR, arg[4], false, lmp);
+  dgradflag = utils::inumeric(FLERR, arg[5], false, lmp);
+  if (dgradflag && !bikflag)
+    error->all(FLERR,"Illegal compute pace command: dgradflag=1 requires bikflag=1");
+
+  // map[i] = which element the Ith atom type is, -1 if not mapped
+  // map[0] is not used
+  memory->create(map,ntypes+1,"pace:map");
+  /*
+  for (int i=0; i<ntypes; i++){
+    map[i+1] = i+1;
```

Comment:
Need to remove commented out code, or make it clear why we it is helpful to keep.